File: leetcode_ly/c++/BackTracking/Permutations.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> elements;
        vector<vector<int>> result;
        if(nums.size() == 0) return result;
        getResult(nums,result,elements);
        return result;
    }
    
    void getResult(vector<int>& nums,vector<vector<int>>& result,vector<int>& elements){
        if(elements.size() == nums.size()){
            result.push_back(elements);
            return;
        }
        for(int i = 0; i < nums.size(); i++){
            if(find(elements.begin(),elements.end(),nums[i]) != elements.end()) continue;
            elements.push_back(nums[i]);
            getResult(nums,result,elements);
            elements.pop_back();
        }
    }
};

File: leetcode_ly/java/Backtracking && DP/Permutations.java
public class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        if(nums == null || nums.length == 0) return res;
        List<Integer> ans = new ArrayList<Integer>();
        getResult(res,nums,ans);
        return res;
    }
    
    public void getResult(List<List<Integer>> res,int[] nums,List<Integer> ans){
        if(ans.size() == nums.length){
            res.add(new ArrayList<>(ans));
            return;
        }
        
        for(int i = 0; i < nums.length; i++){
            if(ans.contains(nums[i])) continue;
            ans.add(nums[i]);
            getResult(res,nums,ans);
            ans.remove(ans.size() - 1);
        }
    }
}